src/protocol/prf_sharding/mod.rs

```diff
@@ -0,0 +1,704 @@
+use std::iter::repeat;
+
+use futures_util::future::try_join_all;
+
+use super::step::BitOpStep;
+use crate::{
+    error::Error,
+    ff::{Field, GaloisField, Gf2},
+    protocol::{
+        basics::{SecureMul, ShareKnownValue},
+        context::{Context, UpgradableContext, UpgradedContext, Validator},
+        BasicProtocols, RecordId,
+    },
+    repeat64str,
+    secret_sharing::{
+        replicated::{semi_honest::AdditiveShare as Replicated, ReplicatedSecretSharing},
+        BitDecomposed, Linear as LinearSecretSharing,
+    },
+};
+
+pub struct PrfShardedIpaInputRow<BK: GaloisField, TV: GaloisField> {
+    prf_of_match_key: u64,
+    is_trigger_bit: Replicated<Gf2>,
+    breakdown_key: Replicated<BK>,
+    trigger_value: Replicated<TV>,
+}
+
+struct InputsRequiredFromPrevRow {
+    ever_encountered_a_source_event: Replicated<Gf2>,
+    attributed_breakdown_key_bits: BitDecomposed<Replicated<Gf2>>,
+    saturating_sum: BitDecomposed<Replicated<Gf2>>,
+    is_saturated: Replicated<Gf2>,
+    difference_to_cap: BitDecomposed<Replicated<Gf2>>,
+}
+
+#[derive(Debug)]
+pub struct CappedAttributionOutputs {
+    pub did_trigger_get_attributed: Replicated<Gf2>,
+    pub attributed_breakdown_key_bits: BitDecomposed<Replicated<Gf2>>,
+    pub capped_attributed_trigger_value: BitDecomposed<Replicated<Gf2>>,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub(crate) enum Step {
+    BinaryValidator,
+    EverEncounteredSourceEvent(usize),
+    DidTriggerGetAttributed(usize),
+    AttributedBreakdownKey(usize),
+    AttributedTriggerValue(usize),
+    ComputeSaturatingSum(usize),
+    IsSaturatedAndPrevRowNotSaturated(usize),
+    ComputeDifferenceToCap(usize),
+    ComputedCappedAttributedTriggerValueNotSaturatedCase(usize),
+    ComputedCappedAttributedTriggerValueJustSaturatedCase(usize),
+}
+
+impl crate::protocol::step::Step for Step {}
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        const EVER_ENCOUNTERED_SOURCE_EVENT: [&str; 64] = repeat64str!["eese_row"];
```

Comment:
This boilerplate code is gone now, thanks to Taiki's recent change. I think you need to mark the step as dynamic and use that instead. Sample example: https://github.com/private-attribution/ipa/blob/main/src/protocol/sort/apply_sort/shuffle.rs#L21